src/silhouette.jl

```diff
@@ -18,48 +84,11 @@ function sil_aggregate_dists(k::Int, a::AbstractVector{Int}, dists::AbstractMatr
     return r
 end
 
-
-"""
-    silhouettes(assignments::AbstractVector, [counts,] dists) -> Vector{Float64}
-    silhouettes(clustering::ClusteringResult, dists) -> Vector{Float64}
-
-Compute *silhouette* values for individual points w.r.t. given clustering.
-
-Returns the ``n``-length vector of silhouette values for each individual point.
-
-# Arguments
- - `assignments::AbstractVector{Int}`: the vector of point assignments
-   (cluster indices)
- - `counts::AbstractVector{Int}`: the optional vector of cluster sizes (how many
-   points assigned to each cluster; should match `assignments`)
- - `clustering::ClusteringResult`: the output of some clustering method
- - `dists::AbstractMatrix`: ``n×n`` matrix of pairwise distances between
-   the points
-
-# References
-> Peter J. Rousseeuw (1987). *Silhouettes: a Graphical Aid to the
-> Interpretation and Validation of Cluster Analysis*. Computational and
-> Applied Mathematics. 20: 53–65.
-"""
-function silhouettes(assignments::AbstractVector{<:Integer},
-                     counts::AbstractVector{<:Integer},
-                     dists::AbstractMatrix{T}) where T<:Real
-
-    n = length(assignments)
-    k = length(counts)
-    k >= 2 || throw(ArgumentError("silhouettes() not defined for the degenerated clustering with a single cluster."))
-    for j = 1:n
-        (1 <= assignments[j] <= k) || throw(ArgumentError("Bad assignments[$j]=$(assignments[j]): should be in 1:$k range."))
-    end
-    sum(counts) == n || throw(ArgumentError("Mismatch between assignments ($n) and counts (sum(counts)=$(sum(counts)))."))
-    size(dists) == (n, n) || throw(DimensionMismatch("The size of a distance matrix ($(size(dists))) doesn't match the length of assignment vector ($n)."))
-
-    # compute average distance from each cluster to each point --> r
-    r = sil_aggregate_dists(k, assignments, dists)
+function normalize_aggregate_dists!(r, assignments, counts, k, n)
     # from sum to average
     @inbounds for j = 1:n
         for i = 1:k
-            c = counts[i]
+            c = copy(counts[i])
```

Comment:
I did this because the variable c is modified during the loop, which I was afraid might modify the counts array in the precomputed variables struct.
I can test without this to see the results are consistent.

Reply:
`Int` is a primitive type, the assignment already copies the bits, not the reference to the array element.

Reply:
You only need to copy if the elements are of mutable type (mutable struct, arrays), or even deepcopy() if the fields are mutable, because the assignment just assigns the reference to the same memory.